### ISSUE TITLE
How can I check that a formula referenced a variable that is not yet ready?

### ISSUE THREAD

**Issue by cesss** [andmarti1424/sc-im#231]
Hi!

I'm trying to understand the order in which operations are performed. There's a dependency graph (in `dep_graph.c`) but it seems this graph doesn't always have the order of operations accurately defined: sometimes the evaluation of a formula is executed, then the evaluation tree detects that a referenced variable is a result from another formula which isn't ready yet, and, in this moment, some flag (the `cellerror` variable perhaps?) is set, so that the final result of the evaluation is discarded.

Later, formulas whose result was discarded, are tried again, and their result is tagged as correct if all the referenced variables are ready this time.

Am I correct in this assumption? (If I'm not correct, then there's a bug in `dep_graph.c` because failed evaluation of formulas caused by non-ready variables happens quite frequently when SC-IM is running).

And now, my question: How can I check (for example inside the `EvalJustOneVertex()` function from `dep_graph.c`) that a formula evaluation succeeded, or needs to be tried again? How can I check in that function that a formula needs to be evaluated again because referenced variables were not ready yet?

Thanks!

**Comment by cesss** [comment]
As a follow up, I detected that it seems there's no way to detect if a variable has been already evaluated or not.

For example, I compared the data in two variables: one of them had been previously evaluated and has a correct value. The other one has not been evaluated yet (so the value is still 0, the default value when a variable has not been evaluated yet).

Both variables have `cellerror=0` and both variables have `flags=259`

So, is it not possible to check if a variable is a formula which has already been evaluated?

**Comment by andmarti1424** [comment]
Vertex.visited indicates if an ent was evaluated.

**Comment by cesss** [comment]
Thanks a lot. However, I sincerely believe something needs to be reconsidered in the way formulas are evaluated. I don't have a test case ready for sending it to you, but my impression is that, at least when using `--nocurses --quit_afterload` (which is my main use scenario for SC-IM), some formulas are evaluated when they shouldn't (before their referenced variables have been evaluated), and even in some cases there're formulas which aren't reevaluated later and they stay with the wrong value of 0.

I understand however that the original SC source code is not a clean design, so this is not your fault, but your task of tidying it up is not easy at all. 

In my honest opinion, there are two design choices in SC (not in SC-IM, but in the original SC) that are serious mistakes and have very bad consequences in how the program behaves, and in maintaining it: The most important is the idea that every cell exists with a default value of 0 and that the check for stopping evaluation is that no cell changed its value in the last evaluation. The consequences of this decision are catastrophic for maintaining the program. If the default value of a cell was a flag indicating "undefined value", then most Eval-related bugs would disappear immediately, because if a formula references an still unevaluated formula, its result would become "undefined value". This would be very easy to maintain.

The second most important mistake (again, IMHO) is separating "numeric formulas" and "string formulas". They should be the same. Syntax errors of numbers vs strings should be checked not when parsing cells, but when evaluating cells. If some operator cannot be applied to the data type of the operand, then issue the syntax error at evaluation time (for example, `@COS("Peter")` or `@UPPER(100+45)`). Of course, the origins of SC predate the years in which object oriented programming became widespread, so the abstraction of operands was not in the mind of the original developers (I'm not saying that they should have used C++: Abstraction of operands data types can be implemented in C without any C++, but you need your mind accustomed to abstraction).

Then, of course, comes the problem of how SC was originally coded, which is quite prone to spaghetti-bugs, and hard to find bugs. But this problem is, in my opinion, less important than the other two I mentioned above. Without the two problems above, maintaining SC-IM would be far easier than it is.

Sorry for my long post, but I needed to say it :-)

**Comment by andmarti1424** [comment]
Yes, bottom up evaluation will be done in next release. This will fix some issues regarding evaluation of formulas. About the numeric /string formulas, you might be right, but it is something I will not change in the immediate future..

**Comment by cesss** [comment]
Thanks a lot. I just faced another scenario were the "every cell has a default value of 0" was the cause for another problem: I added a few code lines for discarding the results of formulas that referenced variables that weren't evaluated yet. It seemed to work fine, until... there was a formula whose correct final value was zero, so:

- My check for discarding formulas that had used not-yet-evaluated variables discarded the first evaluation of the formula.
- And then, the SC check in `EvalJustOneVertex()` that only assigns the `is_changed` flag when the value in the cell has changed, failed, because the value didn't change (it was still zero) no matter if the evaluation was non-legal when it returned zero the last time.

(of course I fixed it by modifying such check in `EvalJustOneVertex()`, so that the `is_changed` flag is also assigned when the discarded condition of the cell has changed).

As I said, the original SC design decision of "every cell has a default value of zero" is evil... **very** evil. If you want to reduce the number of potential bugs in the future, I strongly suggest that you modify the cell `struct` to encapsulate the cell value and hide it from the code. Don't make the value a direct `double` but a `struct value { double val; int undefined; };` so that everytime you access a cell value, you know if it's defined or undefined (when `value.undefined!=0`)

**Comment by andmarti1424** [comment]
That sounds interesting. It could be modded with the bottom up evaluation. You are free to help with pull requests regarding these topics. As i said earlier, the evaluation will be the main change in the next release.

**Comment by cesss** [comment]
I don't have the time for contributing with code right now, but my opinion on how to design future changes in SC-IM could be summarised in my long post above, the one about what I called "mistakes" in the original design of SC.

One clarification, though: when I speak against the decision of cells always having a valid default value of 0, one should differentiate between empty cells, and cells with formulas. A cell which has a formula, should never have a valid "default value", but an "undefined value". On the other hand, empty cells are a different matter: My personal choice would also be to consider their value as "undefined", but this would likely break compatibility with old spreadsheet files that referenced empty cells expecting that their value is 0. For example, Excel permits accessing empty cells and gives 0 as their value, but it usually warns (at least in some version) when a formula accessed empty cells. So it permits it, but warns about it.

In conclusion, I think you should differentiate:

- Cells with formulas should never have a valid "default value". Either they are evaluated and have the correct value, or they have an invalid/undefined value until they are evaluated or re-evaluated.

- Empty cells could be managed with an user preference, allowing a very strict policy (consider that every operation involving an empty cell generates an undefined value), a semi-strict policy (consider empty cells are 0 but warn if they are accessed), or a very permissive policy (consider empty cells are 0 and never warn if they're accessed). I thing it would be very nice to have this as a user preference, as you could use it for debugging that your formulas are not accessing empty cells.

**Comment by andmarti1424** [comment]
Did you update to latest changes? Sc-im is doing a bottom up evaluation of the graph..

**Comment by andmarti1424** [comment]
@cesss ?

**Comment by cesss** [comment]
Sorry @andmarti1424 for not replying: Unfortunately I cannot look at the sc-im code at the moment: I'm using it internally as part of an in-house tool, and I'm using it in a frequent basis, but currently I'm using it as a deployed tool that isn't being modified, only rebuilt.

However, I can give you my advice from my opinion about sc-im (related to the code base I'm using, which is about 2 years old or so): I believe sc-im calls for a reconsideration/redesign of several of its aspects. I'd keep it as C code (please don't complicate it unnecessarily with C++), but however it needs to consider number cells and text cells as equally important citizens. The heavy distinction between numbers and text, is one of the biggest issues in sc-im, IMHO. Also, perhaps the evaluation process must be cleaned up (but you say you modified it, so maybe that part is already improved).

**Comment by cesss** [comment]
Also, related to my opinion about "numbers vs text", take a look at this issue: https://github.com/andmarti1424/sc-im/issues/478

In fact, that issue would be solved if sc-im considered numbers and text transparently, in an abstract way (and you don't need C++ for abstract types: C is perfectly suitable for it).

**Comment by andmarti1424** [comment]
Yes @cesss I dont like it that either. Dont forget this all started tooking sc and modifying it and some addons. I didnt make it from scratch.
My idea for a v2. (if I find colaborators) is to replace its evaluation core and support multi sheet references.
And threat numbers and text one thing. Dont have both on a same cell.

**Comment by andmarti1424** [comment]
I close this. Thanks for your suggestions. Hopefully this would help in the design for v2.
Btw, if interested, please check my latest comments here too: https://github.com/andmarti1424/sc-im/issues/496
